core/controllers/reader_test.py

```diff
@@ -1189,6 +1189,223 @@ def test_remove_collection_from_incomplete_list_handler(self):
+                }
+            },
+            'playthrough_actions': [{
+                'action_type': 'ExplorationStart',
+                'action_customization_args': {
+                    'state_name': {
+                        'value': 'state_name1'
+                    }
+                }
+            }],
+            'is_valid': True
+        }
+
+        self.exp_issue = {
+            'issue_type': 'EarlyQuit',
+            'schema_version': 1,
+            'customization_args': {
+                'state_name': {
+                    'value': 'state_name1'
+                },
+                'time_spent_in_exp_in_msecs': {
+                    'value': 250
+                }
+            }
+        }
+
+        response = self.testapp.get('/explore/%s' % self.exp_id)
+        self.csrf_token = self.get_csrf_token_from_response(response)
+
+    def test_new_playthrough_gets_stored(self):
+        """Test that a new playthrough gets created and is added to an existing
+        issue's list of playthrough IDs.
+        """
+        self.post_json(
+            '/explorehandler/store_playthrough/%s' % (self.exp_id),
+            {
+                'playthrough_data': self.playthrough_data,
+                'exp_issue': self.exp_issue
+            }, self.csrf_token)
+        self.process_and_flush_pending_tasks()
+
+        model = stats_models.ExplorationIssuesModel.get(self.exp_id)
+        self.assertEqual(len(model.unresolved_issues), 1)
+        self.assertEqual(len(model.unresolved_issues[0]['playthrough_ids']), 2)
+
+    def test_new_exp_issue_gets_created(self):
+        """Test that a new playthrough gets created and a new issue is created
+        for it.
+        """
+        self.exp_issue['customization_args']['state_name']['value'] = (
+            'state_name2')
+        self.playthrough_data['issue_customization_args']['state_name'][
+            'value'] = 'state_name2'
+
+        self.post_json(
+            '/explorehandler/store_playthrough/%s' % (self.exp_id),
+            {
+                'playthrough_data': self.playthrough_data,
+                'exp_issue': self.exp_issue
+            }, self.csrf_token)
+        self.process_and_flush_pending_tasks()
+
+        model = stats_models.ExplorationIssuesModel.get(self.exp_id)
+        self.assertEqual(len(model.unresolved_issues), 2)
+        self.assertEqual(len(model.unresolved_issues[0]['playthrough_ids']), 1)
+        self.assertEqual(len(model.unresolved_issues[1]['playthrough_ids']), 1)
+
+    def test_playthrough_gets_added_to_cyclic_issues(self):
+        """Test that a new cyclic playthrough gets added to the correct
+        cyclic issue when it exists.
+        """
+        playthrough_id = stats_models.PlaythroughModel.create(
+            self.exp_id, self.exploration.version, 'CyclicStateTransitions',
+            {
+                'state_names': {
+                    'value': ['state_name1', 'state_name2', 'state_name1']
```

Comment:
Any checks that playthroughs with the same elements in this list but in different orders are treated differently? Should have a test for that.

Reply:
Yep. Just added one.